File: p1799_Maximize_Score_After_N_Operations/submission_recursive.cpp
bool const debug = false;
int const INVALID = -1;
int const MAXN = 14;
int cache[(MAXN >> 1)+1][(1 << MAXN)];

/*
test cases
[1,2]
[3,4,6,8]
[1,2,3,4,5,6]
[39759,619273,859218,228161,944571,597983,483239,179849,868130,909935,912143,817908,738222,653224]
[481851,31842,817070,452937,627635,712245]
*/
class Solution {
private:
    int gcd(int big, int small) {
        if (big < small) swap(big, small);
        if (small == 0) return big;
        return gcd(small, big%small);
    }

    int dfs(int fixedU, int step, vector<unordered_map<int, int>>& adj, int remainedBm) {
        int m = adj.size(), ret = 0;
        if (INVALID == fixedU) {
            if (INVALID != cache[step][remainedBm]) return cache[step][remainedBm];
            for (int u = 0; u < m; ++u) {
                int mask = (1 << u);
                if (false == (remainedBm & mask)) continue;
                if (debug) printf("%*sStep:%d, Checking {remainedBm:%d, u:%d}\n", (2*(m/2-step)), "", step, remainedBm, u);
                int cand = dfs(u, step, adj, (remainedBm^mask)); 
                ret = max(ret, cand);
            }
            cache[step][remainedBm] = ret;
        } else {
            for (auto &[v, vWeight] : adj[fixedU]) {
                int mask = (1 << v);
                if (false == (remainedBm & mask)) continue;
                if (debug) printf("%*sStep:%d, Checking {remainedBm:%d, u:%d, v:%d, gcdVal:%d}\n", (2*(m/2-step)), "", step, remainedBm, fixedU, v, vWeight);
                int cand = vWeight*step + dfs(INVALID, step-1, adj, (remainedBm^mask)); 
                if (cand > ret) {
                    ret = cand;
                }
            }
        }
        return ret;
    }

public:
    int maxScore(vector<int>& nums) {
        memset(cache, INVALID, sizeof cache);
        int m = nums.size();
        vector<unordered_map<int, int>> adj(m);
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < m; ++j) {
                if (j == i) continue;
                int gcdVal = gcd(nums[j], nums[i]);
                adj[i][j] = gcdVal;
                adj[j][i] = gcdVal;
            }
        }

        int remainedBm = (1 << m) - 1, step = (m >> 1);
        int ans = dfs(INVALID, step, adj, remainedBm);
        return ans;
    }
};

File: p996_Number_of_Squareful_Arrays/submission_recursive.py
debug = True

def isSquare(val):
    return math.isqrt(val) ** 2 == val
    
def buildAdj(cnt):
    adj = defaultdict(set)
    for k1 in cnt.keys():            
        for k2 in cnt.keys():
            if isSquare(k1+k2):
                adj[k1].add(k2)
    return adj
    
def dfs(buff, n, cnt, adj):
    if len(buff) == n:
        return 1
    
    res = 0
    if not buff:
        for k in cnt.keys():
            buff.append(k)
            cnt[k] -=1 
            res += dfs(buff, n, cnt, adj)
            cnt[k] +=1 # backtracking
            buff.pop() # backtracking
    else:
        for k in adj[buff[-1]]:
            if 0 == cnt[k]:
                continue
            buff.append(k)
            cnt[k] -=1
            res += dfs(buff, n, cnt, adj)
            cnt[k] +=1 # backtracking
            buff.pop() # backtracking
    return res
    
"""
test cases
[1,17,8]
[2,2,2]
[2,2,2,2,2,2,2,2,2,2,2,2]
"""
class Solution:
    def numSquarefulPerms(self, nums: List[int]) -> int:
        n = len(nums)
        cnt = defaultdict(int)
        for val in nums:
            cnt[val] += 1
        
        adj = buildAdj(cnt)
        buff = []
        ans = dfs(buff, n, cnt, adj)
        return ans
